File: app/src/main/java/com/tangbba/simplegithubforjava/ui/search/SearchActivity.java
package com.tangbba.simplegithubforjava.ui.search;

import android.content.Intent;
import android.support.v7.app.ActionBar;
import android.support.v7.app.AppCompatActivity;
import android.os.Bundle;
import android.support.v7.widget.LinearLayoutManager;
import android.support.v7.widget.RecyclerView;
import android.support.v7.widget.SearchView;
import android.view.Menu;
import android.view.MenuItem;
import android.view.View;
import android.view.inputmethod.InputMethodManager;
import android.widget.ProgressBar;
import android.widget.TextView;

import com.tangbba.simplegithubforjava.R;
import com.tangbba.simplegithubforjava.api.GithubApi;
import com.tangbba.simplegithubforjava.api.GithubApiProvider;
import com.tangbba.simplegithubforjava.api.model.GithubRepo;
import com.tangbba.simplegithubforjava.api.model.RepoSearchResponse;
import com.tangbba.simplegithubforjava.ui.repo.RepositoryActivity;

import retrofit2.Call;
import retrofit2.Callback;
import retrofit2.Response;

public class SearchActivity extends AppCompatActivity implements SearchAdapter.ItemClickListener {

    RecyclerView mRecyclerView;
    ProgressBar mProgressBar;
    TextView mSearchMessageTextView;
    MenuItem mSearchMenu;
    SearchView mSearchView;
    SearchAdapter mAdapter;

    GithubApi mGithubApi;
    Call<RepoSearchResponse> mRepoSearchResponseCall;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_search);

        mRecyclerView = findViewById(R.id.search_result_recycler_view);
        mProgressBar = findViewById(R.id.progress_bar);
        mSearchMessageTextView = findViewById(R.id.search_message_text_view);

        mAdapter = new SearchAdapter();
        mAdapter.setItemClickListener(this);
        mRecyclerView.setLayoutManager(new LinearLayoutManager(this));
        mRecyclerView.setAdapter(mAdapter);

        mGithubApi = GithubApiProvider.provideGithubApi(this);
    }

    @Override
    public boolean onCreateOptionsMenu(Menu menu) {
        getMenuInflater().inflate(R.menu.menu_activity_search, menu);
        mSearchMenu = menu.findItem(R.id.menu_activity_search_query);

        mSearchView = (SearchView) mSearchMenu.getActionView();
        mSearchView.setOnQueryTextListener(new SearchView.OnQueryTextListener() {
            @Override
            public boolean onQueryTextSubmit(String query) {
                updateTitle(query);
                hideSoftKeyboard();
                collapseSearchView();
                searchRepository(query);
                return true;
            }

            @Override
            public boolean onQueryTextChange(String newText) {
                return false;
            }
        });

        mSearchMenu.expandActionView();
        return true;
    }

    @Override
    public boolean onOptionsItemSelected(MenuItem item) {
        if (R.id.menu_activity_search_query == item.getItemId()) {
            item.expandActionView();
            return true;
        }
        return super.onOptionsItemSelected(item);
    }

    @Override
    public void onItemClick(GithubRepo repository) {
        Intent intent = new Intent(this, RepositoryActivity.class);
        intent.putExtra(RepositoryActivity.KEY_USER_LOGIN, repository.getOwner().getLogin());
        intent.putExtra(RepositoryActivity.KEY_REPO_NAME, repository.getName());
        startActivity(intent);
    }

    private void searchRepository(String query) {
        clearResults();
        hideError();
        showProgress();

        mRepoSearchResponseCall = mGithubApi.searchRepository(query);
        mRepoSearchResponseCall.enqueue(new Callback<RepoSearchResponse>() {
            @Override
            public void onResponse(Call<RepoSearchResponse> call,
                                   Response<RepoSearchResponse> response) {
                hideProgress();
                RepoSearchResponse result = response.body();
                if (response.isSuccessful() && null != response) {
                    mAdapter.setItems(result.getItems());
                    mAdapter.notifyDataSetChanged();

                    if (0 == result.getTotalCount()) {
                        showError(getString(R.string.no_search_result));
                    }
                } else {
                    showError("Not successful: " + response.message());
                }
            }

            @Override
            public void onFailure(Call<RepoSearchResponse> call, Throwable t) {
                hideProgress();
                showError(t.getMessage());
            }
        });
    }

    private void updateTitle(String query) {
        ActionBar actionBar = getSupportActionBar();
        if (null != actionBar) {
            actionBar.setSubtitle(query);
        }
    }

    private void hideSoftKeyboard() {
        InputMethodManager imm = (InputMethodManager) getSystemService(INPUT_METHOD_SERVICE);
        imm.hideSoftInputFromWindow(mSearchView.getWindowToken(), 0);
    }

    private void collapseSearchView() {
        mSearchMenu.collapseActionView();
    }

    private void clearResults() {
        mAdapter.clearItems();
        mAdapter.notifyDataSetChanged();
    }

    private void showProgress() {
        mProgressBar.setVisibility(View.VISIBLE);
    }

    private void hideProgress() {
        mProgressBar.setVisibility(View.GONE);
    }

    private void showError(String message) {
        mSearchMessageTextView.setText(message);
        mSearchMessageTextView.setVisibility(View.VISIBLE);
    }

    private void hideError() {
        mSearchMessageTextView.setText("");
        mSearchMessageTextView.setVisibility(View.GONE);
    }
}

File: app/src/main/kotlin/com/tangbba/simplegithubforjava/ui/search/SearchActivity.kt
package com.tangbba.simplegithubforjava.ui.search

import android.content.Context
import android.content.Intent
import android.os.Bundle
import android.support.v7.app.AppCompatActivity
import android.support.v7.widget.LinearLayoutManager
import android.support.v7.widget.SearchView
import android.view.Menu
import android.view.MenuItem
import android.view.View
import android.view.inputmethod.InputMethodManager
import com.tangbba.simplegithubforjava.R
import com.tangbba.simplegithubforjava.api.GithubApi
import com.tangbba.simplegithubforjava.api.model.GithubRepo
import com.tangbba.simplegithubforjava.api.model.RepoSearchResponse
import com.tangbba.simplegithubforjava.api.provideGithubApi
import com.tangbba.simplegithubforjava.ui.repo.RepositoryActivity
import kotlinx.android.synthetic.main.activity_search.*
import org.jetbrains.anko.startActivity
import retrofit2.Call
import retrofit2.Callback
import retrofit2.Response

class SearchActivity : AppCompatActivity(), SearchAdapter.ItemClickListener {

    internal lateinit var mSearchMenu: MenuItem
    internal lateinit var mSearchView: SearchView
    internal val mAdapter: SearchAdapter by lazy {
        SearchAdapter().apply { setItemClickListener(this@SearchActivity) }
    }

    internal val mGithubApi: GithubApi by lazy {
        provideGithubApi(this)
    }

    internal var mRepoSearchResponseCall: Call<RepoSearchResponse>? = null

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_search)

        with(search_result_recycler_view) {
            layoutManager = LinearLayoutManager(this@SearchActivity)
            adapter =this@SearchActivity.mAdapter
        }
    }

    override fun onStop() {
        super.onStop()
        mRepoSearchResponseCall?.run { cancel() }
    }

    override fun onCreateOptionsMenu(menu: Menu): Boolean {
        menuInflater.inflate(R.menu.menu_activity_search, menu)
        mSearchMenu = menu.findItem(R.id.menu_activity_search_query)

        mSearchView = (mSearchMenu.actionView as SearchView).apply {
            setOnQueryTextListener(object : SearchView.OnQueryTextListener {
                override fun onQueryTextSubmit(query: String): Boolean {
                    updateTitle(query)
                    hideSoftKeyboard()
                    collapseSearchView()
                    searchRepository(query)
                    return true
                }

                override fun onQueryTextChange(newText: String?): Boolean {
                    return false
                }

            })
        }

        with(mSearchMenu) {
            setOnActionExpandListener(object : MenuItem.OnActionExpandListener {
                override fun onMenuItemActionExpand(item: MenuItem?): Boolean {
                    return true
                }

                override fun onMenuItemActionCollapse(item: MenuItem?): Boolean {
                    if ("" == mSearchView.query) {
                        finish()
                    }
                    return true
                }

            })
            expandActionView()
        }
        return true
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        if (R.id.menu_activity_search_query == item.itemId) {
            item.expandActionView()
            return true
        }
        return super.onOptionsItemSelected(item)
    }

    override fun onItemClick(repository: GithubRepo) {
        startActivity<RepositoryActivity>(
                RepositoryActivity.KEY_USER_LOGIN to repository.owner.login,
                RepositoryActivity.KEY_REPO_NAME to repository.name
        )
    }

    private fun searchRepository(query: String) {
        clearResults()
        hideError()
        showProgress()

        mRepoSearchResponseCall = mGithubApi.searchRepository(query)
        mRepoSearchResponseCall!!.enqueue(object : Callback<RepoSearchResponse> {
            override fun onResponse(call: Call<RepoSearchResponse>,
                                    response: Response<RepoSearchResponse>?) {
                hideProgress()

                val result = response!!.body()
                if (response!!.isSuccessful && null != response) {
                    with(mAdapter) {
                        setItems(result!!.items)
                        notifyDataSetChanged()
                    }

                    if (0 == result!!.totalCount) {
                        showError(getString(R.string.no_search_result))
                    }
                } else {
                    showError("Not successful: " + response.message())
                }
            }

            override fun onFailure(call: Call<RepoSearchResponse>, t: Throwable) {
                hideProgress()
                showError(t.message)
            }
        })
    }

    private fun updateTitle(query: String) {
        supportActionBar?.run {
            subtitle = query
        }
    }

    private fun hideSoftKeyboard() {
        (getSystemService(Context.INPUT_METHOD_SERVICE) as InputMethodManager).run {
            hideSoftInputFromWindow(mSearchView.windowToken, 0)
        }
    }

    private fun collapseSearchView() {
        mSearchMenu.collapseActionView()
    }

    private fun clearResults() {
        with(mAdapter) {
            clearItems()
            notifyDataSetChanged()
        }
    }

    private fun showProgress() {
        progress_bar.visibility = View.VISIBLE
    }

    private fun hideProgress() {
        progress_bar.visibility = View.GONE
    }

    private fun showError(message: String?) {
        with(search_message_text_view) {
            text = message ?: "Unexpected error."
            visibility = View.VISIBLE
        }
    }

    private fun hideError() {
        with(search_message_text_view) {
            text = ""
            visibility = View.GONE
        }
    }
}
